File: sample_source_code/python/src/cat.py
"""Cat: concrete subclass of Animal with lifespan estimation."""
from __future__ import annotations

from animal import Animal
from math_utils import MathUtils


class Cat(Animal):
    """Cat with indoor/outdoor lifespan calculation."""

    def __init__(self, name: str, age: int, is_indoor: bool, weight: float) -> None:
        super().__init__(name, age, "Felis catus", weight)
        self.is_indoor = is_indoor

    def sound(self) -> str:
        return "Meow"

    def accept(self, visitor) -> None:
        visitor.visit_cat(self)

    def expected_lifespan(self) -> int:
        """Returns expected lifespan adjusted for environment and weight."""
        base       = 15 if self.is_indoor else 10
        deviation  = abs(self._weight - 4.5) / 4.5
        adjustment = int(MathUtils.linear_decay(deviation, 1.0, 5))
        return max(5, base - adjustment)

File: sample_source_code/javascript/src/Cat.js
const Animal    = require('./Animal');
const MathUtils = require('./MathUtils');

/**
 * Concrete Cat with indoor/outdoor lifespan estimation.
 * @extends Animal
 */
class Cat extends Animal {
    /**
     * @param {string}  name
     * @param {number}  age
     * @param {boolean} isIndoor
     * @param {number}  weight
     */
    constructor(name, age, isIndoor, weight) {
        super(name, age, 'Felis catus', weight);
        this.isIndoor = isIndoor;
    }

    sound()         { return 'Meow'; }
    accept(visitor) { visitor.visitCat(this); }

    /**
     * Expected lifespan in years based on environment and weight.
     * @returns {number}
     */
    expectedLifespan() {
        const base       = this.isIndoor ? 15 : 10;
        const deviation  = Math.abs(this._weight - 4.5) / 4.5;
        const adjustment = Math.floor(MathUtils.linearDecay(deviation, 1.0, 5));
        return Math.max(5, base - adjustment);
    }
}

module.exports = Cat;
